Bump the Go version

mattfarina opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:
Use a supported version of Go.

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

mattfarina committed:
Bump the Go version

Needed to gofmt source to meet changes in style

Signed-off-by: Matt Farina <matt.farina@suse.com>

TerryHowe commented on .github/workflows/build-test.yml:
```diff
@@ -26,8 +26,8 @@ jobs:
         env:
-          GOLANGCI_LINT_VERSION: '1.46.2'
-          GOLANGCI_LINT_SHA256: '242cd4f2d6ac0556e315192e8555784d13da5d1874e51304711570769c4f2b9b'
+          GOLANGCI_LINT_VERSION: '1.51.2'
```
This version works on mac! 1.50 with go 1.20 was borked.

hickeyma reviewed (changes requested):
Looks good @mattfarina except should the `go.mod` be bumped to 1.20?

mattfarina committed:
Updating the Go version in go.mod

At this time both Go 1.19 and 1.20 are supported. The version
specified in the go.mod file is the minimum version we expect Helm
to be compiled against. This is the oldest supported version to
support environments where others compile Helm. The Helm project
is using Go 1.20 to build Helm itself.

Updating to Go 1.19 also includes dealing with io/ioutil
deprecation and some additional linting issues around staticcheck.
All the staticcheck issues were in test files so linting was
skipped for those.

Signed-off-by: Matt Farina <matt.farina@suse.com>

scottrigby commented:
@mattfarina @souleb  @sabre1041 @jdolitsky hmm Matt I see your test details in Keybase, fun change in Go 🧐 If we don't have a solution for this now I agree we should revert in the meantime, work to resolve this and then bring back into main after resolving (pun intended!). To help us with this - and for future, any objections to setting up our GitHub action on a matrix of Helm's supported Go versions, architectures, etc? https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs

scottrigby commented:
@souleb is working on a potential fix. @mattfarina any objection to holding off on reverting for a bit? Say until Monday?

mattfarina committed:
Handle failed DNS case for Go 1.20+

Go 1.20 introduced DNS/CNAME handling changes. This can cause an
IP address represented in hex, oct, etc to be looked up as DNS and
fail. This change introduces a mock DNS resolver.

Note, with the mock resolver, we don't need to use 0x7f000001 any
longer. Keeping because it was already there.

Signed-off-by: Matt Farina <matt.farina@suse.com>

mattfarina commented:
@hickeyma can you please re-review? It's working now.

The large size of the change is because of the deprecation of ioutil and the linter making it an error now. The change has been around for that long.

Also, there is a mock DNS resolver because some systems have trouble with the system resolvers which are inconsistent.

hickeyma reviewed (approved):
Great work on this @mattfarina especially tracking and fixing the dns issue.
